processor/awscontainerinsightprocessor/processor_test.go

```diff
@@ -0,0 +1,140 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awscontainerinsightprocessor
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+	"testing"
+	"time"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+type MockStore struct {
+	RefreshCount int
+	wg           sync.WaitGroup
+}
+
+func (m *MockStore) Decorate(metric pdata.ResourceMetrics, kubernetesBlob map[string]interface{}) bool {
+	return true
+}
+
+func (m *MockStore) RefreshTick(ctx context.Context) {
+	m.RefreshCount++
+	if m.RefreshCount <= 2 {
+		m.wg.Done()
+	}
+}
+
+func (m *MockStore) currentRefreshCount() int {
+	m.wg.Wait()
+	return m.RefreshCount
+}
+
+type LongRunningMockStore struct {
+	RefreshCount int
+}
+
+func (m *LongRunningMockStore) Decorate(metric pdata.ResourceMetrics, kubernetesBlob map[string]interface{}) bool {
+	return true
+}
+
+func (m *LongRunningMockStore) RefreshTick(ctx context.Context) {
+	m.RefreshCount++
+
+	if m.RefreshCount == 2 {
+		wait := make(chan bool)
+		//to simulate a long running store
+		go func() {
+			fmt.Println("before the long running refresh")
+			time.Sleep(20 * time.Second)
+			close(wait)
+			fmt.Println("after the long running refresh")
+		}()
+		select {
+		case <-wait:
+			fmt.Println("finishied the long running refresh")
+		case <-ctx.Done():
+			fmt.Println("the passed context times out")
+			return
+		}
+	}
+}
+
+func TestStartAndShutdown(t *testing.T) {
+	os.Setenv("HOST_NAME", "host_name")
+	os.Setenv("HOST_IP", "host_ip")
+
+	processor := &awscontainerinsightprocessor{
+		storeRefreshMinInterval: time.Millisecond,
+		storeRefreshTimeout:     2 * time.Millisecond,
+		storeRefreshStoppedC:    make(chan bool),
+	}
+	mockStore := &MockStore{}
+	mockStore.wg.Add(2)
+	processor.stores = append(processor.stores, mockStore)
+	processor.start()
+
+	//block until refreshTick() executes twice
+	if mockStore.currentRefreshCount() != 2 {
+		t.Fail()
+	}
+
+	processor.Shutdown()
+	//wait two refresh intervals for the channel to fire
+	time.Sleep(2 * time.Millisecond)
+
+	select {
+	case <-processor.storeRefreshStoppedC:
+	default:
+		t.Error("Channel is not closed")
+	}
+
+	os.Unsetenv("HOST_NAME")
+	os.Unsetenv("HOST_IP")
+}
+
+func TestStartAndShutdownWithLongRunningStore(t *testing.T) {
+	os.Setenv("HOST_NAME", "host_name")
+	os.Setenv("HOST_IP", "host_ip")
+
+	processor := &awscontainerinsightprocessor{
+		storeRefreshMinInterval: time.Millisecond,
+		storeRefreshTimeout:     2 * time.Millisecond,
+		storeRefreshStoppedC:    make(chan bool),
+	}
+	mockStore := &LongRunningMockStore{}
+	processor.stores = append(processor.stores, mockStore)
+	processor.start()
+
+	//wait 4 refresh interval for the refreshTick() to get stuck on second call
+	time.Sleep(4 * time.Millisecond)
+
+	processor.Shutdown()
+	//wait for the channel to fire
+	time.Sleep(2 * time.Millisecond)
```

Comment:
I expect that these are flaky tests waiting to happen.  Can they be restructured to behave in a deterministic manner regardless how the runtime decides to schedule goroutines?